tests/e2e/json_utils.go

```diff
@@ -0,0 +1,412 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// MarshalJSON marshals a step into JSON while including the type of the action.
+func (step Step) MarshalJSON() ([]byte, error) {
+	actionType := reflect.TypeOf(step.Action)
+
+	result := struct {
+		ActionType string
+		Action     interface{}
+		State      State
+	}{
+		ActionType: actionType.String(),
+		Action:     step.Action,
+		State:      step.State,
+	}
+
+	return json.Marshal(result)
+}
+
+// UnmarshalJSON unmarshals a step from JSON while including the type of the action.
+func (step *Step) UnmarshalJSON(data []byte) error {
+	var tmp struct {
+		ActionType string
+		Action     json.RawMessage
+		State      State
+	}
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+
+	action, err := UnmarshalMapToActionType(tmp.Action, tmp.ActionType)
+	if err != nil {
+		return err
+	}
+
+	step.Action = action
+	step.State = tmp.State
+	return nil
+}
+
+// UnmarshalMapToActionType takes a JSON object and an action type and marshals into an object of the corresponding action.
+func UnmarshalMapToActionType(rawAction json.RawMessage, actionTypeString string) (interface{}, error) {
+	var err error
+	switch actionTypeString {
+	case "main.submitConsumerAdditionProposalAction":
```

Comment:
I'm probably looking to an older local codebase on all the actions we have. I couldn't find any missing. To contrary, the following ones seem new:
```
main.submitParamChangeLegacyProposalAction
main.LegacyUpgradeProposalAction
main.waitUntilBlockAction
main.StartSovereignChainAction
main.lightClientEquivocationAttackAction
main.lightClientAmnesiaAttackAction
main.lightClientLunaticAttackAction
```

Reply:
Yes, that's an old codebase. This should be up-to-date with the actions here:
https://github.com/cosmos/interchain-security/blob/55d2686593b33384b065937ed5ed9ee5304f0536/tests/e2e/main.go#L193